internal/observability/call.go

```diff
@@ -75,69 +75,67 @@ type call struct {
 	levels *levels
 }
 
+type callResult struct {
+	err            error
+	ctxOverrideErr error
+
+	isApplicationError   bool
+	applicationErrorMeta *transport.ApplicationErrorMeta
+
+	requestSize  int
+	responseSize int
+}
+
 type levels struct {
 	success, failure, applicationError zapcore.Level
 }
 
-func (c call) End(err error, requestSize int) {
-	c.endWithAppError(err, false /* isApplicationError */, nil /* applicationErrorMeta */, requestSize, 0)
+func (c call) End(res *callResult) {
+	c.endWithAppError(res)
 }
 
-func (c call) EndCallWithAppError(
-	err error,
-	isApplicationError bool,
-	applicationErrorMeta *transport.ApplicationErrorMeta,
-) {
-	c.endWithAppError(err, isApplicationError, applicationErrorMeta, 0, 0)
+func (c call) EndCallWithAppError(res *callResult) {
+	c.endWithAppError(res)
 }
 
-func (c call) EndHandleWithAppError(err error,
-	isApplicationError bool,
-	applicationErrorMeta *transport.ApplicationErrorMeta,
-	ctxOverrideErr error,
-	requestSize int,
-	responseSize int) {
-	if ctxOverrideErr == nil {
-		c.endWithAppError(err, isApplicationError, applicationErrorMeta, requestSize, responseSize)
+func (c call) EndHandleWithAppError(res *callResult) {
+	if res.ctxOverrideErr == nil {
+		c.endWithAppError(res)
 		return
 	}
 
 	// We'll override the user's response with the appropriate context error. Also, log
 	// the dropped response.
 	var droppedField zap.Field
-	if isApplicationError && err == nil { // Thrift exceptions
+	if res.isApplicationError && res.err == nil { // Thrift exceptions
 		droppedField = zap.String(_dropped, _droppedAppErrLog)
-	} else if err != nil { // other errors
-		droppedField = zap.String(_dropped, fmt.Sprintf(_droppedErrLogFmt, err))
+	} else if res.err != nil { // other errors
+		droppedField = zap.String(_dropped, fmt.Sprintf(_droppedErrLogFmt, res.err))
 	} else {
 		droppedField = zap.String(_dropped, _droppedSuccessLog)
 	}
 
-	c.endWithAppError(ctxOverrideErr,
-		false, /* application error */
-		nil,   /* application failure */
-		requestSize,
-		responseSize,
+	c.endWithAppError(&callResult{
+		err:          res.ctxOverrideErr,
+		requestSize:  res.requestSize,
+		responseSize: res.responseSize,
+	},
```

Comment:
you can use `res` here, no need of recreating this struct

Reply:
CallResult needs to be cleaned up to set `isApplicationError`, `ctxOverrideErr`, and others as nil/empty.
Also, `err` is being overwritten by `ctxOverrideError`. Creating new callResult seems to be the easiest.